cpp/ql/lib/semmle/code/cpp/dataflow/internal/DataFlowImplCommon.qll

```diff
@@ -71,25 +71,31 @@ predicate accessPathCostLimits(int apLimit, int tupleLimit) {
  * calls. For this reason, we cannot reuse the code from `DataFlowImpl.qll` directly.
  */
 private module LambdaFlow {
-  private predicate viableParamNonLambda(DataFlowCall call, int i, ParamNode p) {
-    p.isParameterOf(viableCallable(call), i)
+  private predicate viableParamNonLambda(DataFlowCall call, ArgumentPosition apos, ParamNode p) {
+    exists(ParameterPosition ppos |
+      p.isParameterOf(viableCallable(call), ppos) and
+      parameterMatch(ppos, apos)
```

Comment:
I'm not sure this is the right join-order, when `parameterMatch` isn't the identity.
I'd expect the following to do better:
```
private predicate argumentPosition(DataFlowCall call, ArgNode arg, ParameterPosition ppos) {
  exists(ArgumentPosition apos |
    arg.argumentOf(call, apos) and
    parameterMatch(ppos, apos)
  )
}
private predicate viableParamNonLambda(DataFlowCall call, ParameterPosition ppos, ParamNode p) {
  p.isParameterOf(viableCallable(call), ppos)
}
private predicate viableParamArgNonLambda(DataFlowCall call, ParamNode p, ArgNode arg) {
  exists(ParameterPosition ppos |
    viableParamNonLambda(call, ppos, p) and
    argumentPosition(call, arg, ppos)
  )
}
```

Reply:
I agree that this i more safe; I have pushed updates.